src/muse/costs.py

```diff
@@ -1,33 +1,245 @@
+inputs). Prices, production and consumption data should be split across timeslices
+(i.e. have a "timeslice" dimension). Technology parameters may also be specified at
+the timeslice level, but capacity should not be.
+
+`timeslice_level` is given as an argument to some functions, but in this case it must
+match the timeslice level of any timesliced inputs, so is generally not configurable,
+but only included to ensure consistency and provide explicitness.
+
+The `technologies` input will usually contain data for multiple technologies and have
+a "technology" dimension (sometimes called "asset" or "replacement"). In this case,
+it's important that the `capacity`, `production`, and `consumption` inputs have a
+similar dimension to ensure that costs are calculated for all technologies and to
+prevent unwanted broadcasting.
+
+Additional dimensions (such as "region") may be present in the inputs, but it's up to
+the parent functions to ensure that these are consistent between inputs to prevent
+unwanted broadcasting.
+
+The dimensions of the output will be the sum of all dimensions from the input data,
+minus "commodity", plus "timeslice" (if not already present).
+
+Some functions have a `method` argument, which can be "annual" or "lifetime":
+
+Costs can either be annual or lifetime:
+- annual: calculates the cost for each timeslice in a single year
+- lifetime: calculates the total cost in each timeslice over the lifetime of the
+    technology, using the `technical_life` attribute from the `technologies` dataset.
+    - In this case, technology parameters, production, consumption, capacity and prices
+        are assumed to be constant over the lifetime of the technology. The cost in each
+        year is discounted according to the `interest_rate` attribute from the
+        `technologies` dataset, and summed across years.
+    - Capital costs are different, as these are a one time cost for the lifetime of the
+        technology. This can be annualized by dividing by the `technical_life`.
+Some functions can calculate both lifetime and annual costs, with a `method` argument
+to specify. Others can only calculate one or the other (see individual function
+docstrings for more details).
+
 """
 
-from typing import Optional, Union
+from __future__ import annotations
+
+from functools import wraps
 
 import numpy as np
 import xarray as xr
 
 from muse.commodities import is_enduse, is_fuel, is_material, is_pollutant
-from muse.quantities import consumption
+from muse.quantities import production_amplitude
 from muse.timeslices import broadcast_timeslice, distribute_timeslice
 from muse.utilities import filter_input
 
 
+def cost(func):
+    """Decorator to validate the output dimensions of the cost functions."""
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        result = func(*args, **kwargs)
+        assert "year" not in result.dims
+        assert "timeslice" in result.dims
```

Comment:
Following the description in the docstring above, maybe you could check that `commodity` is not a dimension, either.